src/core/packages/notifications/browser-internal/src/notification_coordinator.ts

```diff
@@ -0,0 +1,124 @@
+   * a lock for the registered entity and buffer new values to be emitted when the provided condition is not met.
+   * Also when the passed observable has been emptied if the current lock still belongs to the registrar it will be released automatically, however an acquired lock can
+   * also be released as deemed fit using {@link releaseLock}.
+   * @param registrar - A unique identifier for the caller of this method
+   * @param $ - Observable to be controlled
+   * @param cond - Condition under which updates from the provided observable should be emitted
+   */
+  public optInToCoordination<T extends Array<{ id: string }>>(
+    registrar: string,
+    $: Observable<T>,
+    cond: Parameters<typeof filter<NotificationCoordinatorState>>[0]
+  ) {
+    // signal used to determine when to emit values from the source observable based on the provided opt-in condition
+    const on$ = this.coordinationLock$.pipe(filter((...args) => cond(...args)));
+    // signal used to determine when to buffer values from the source observable based on the provided opt-in condition
+    const off$ = this.coordinationLock$.pipe(filter((...args) => !cond(...args)));
+    const multicast$ = $.pipe(share());
+
+    return merge(
+      multicast$.pipe(bufferToggle(off$, () => on$)),
+      multicast$.pipe(windowToggle(on$, () => off$))
+    ).pipe(
+      mergeMap((x) => x),
+      tap((value) => {
+        const lock = this.coordinationLock$.getValue();
+        // if the source has values to emit and the lock is not owned (because of values that were buffered whilst the lock was owned by another registrar) we acquire a lock for said source.
+        if (value.length && !lock.locked) {
+          this.acquireLock(registrar);
+        } else if (!value.length && lock.controller === registrar) {
+          // here we handle the scenario where the source observable has been emptied out,
+          // in such event if the lock is still held by the registrar we release it
+          this.releaseLock(registrar);
+        }
+      }),
+      finalize(() => {
+        // when the coordinated observable is completed we release the lock if it is still held by the registrar
+        const lock = this.coordinationLock$.getValue();
+        if (lock.locked && lock.controller === registrar) {
+          this.releaseLock(registrar);
+        }
+      })
+    );
+  }
+}
+
+export function notificationCoordinator(
+  this: Coordinator,
+  registrar: string
+): NotificationCoordinatorPublicApi {
+  return {
+    optInToCoordination: <T extends Array<{ id: string }>>(
+      ...args: Parameters<typeof this.optInToCoordination<T>> extends [infer Head, ...infer Tail]
+        ? Tail
+        : unknown
+    ) =>
+      this.optInToCoordination.apply<
+        Coordinator,
+        Parameters<typeof this.optInToCoordination<T>>,
+        Observable<T>
+      >(this, [registrar, ...args]),
```

Comment:
For consistency with `src/core/packages/notifications/browser/src/types.ts`:

```suggestion
    optInToCoordination: <T extends Array<{ id: string }>>(
      $: Observable<T>,
      cond: (coordinatorState: NotificationCoordinatorState) => boolean
    ) =>
      this.optInToCoordination.apply<
        Coordinator,
        Parameters<typeof this.optInToCoordination<T>>,
        Observable<T>
      >(this, [registrar, $, cond]),
```